text/0565-deprecation-shaking.md

```diff
@@ -0,0 +1,190 @@
+guarantees.
+
+This is especially true now that with Octane, which includes a large new set of
+features that were ultimately designed to completely replace lots of older
+features. Classic features aren't deprecated yet, but at some point in the
+future they likely will be, and being able to take advantage of those size
+reductions as they come would be a huge selling point for Ember apps.
+
+## Detailed design
+
+This RFC proposes allowing users to specify deprecation compliance for both apps
+and addons via a field in the `ember` property in `package.json`:
+
+```json
+{
+  "ember": {
+    "edition": "octane",
+    "deprecationCompliant": {
+      "ember-source": "3.15.0"
+    }
+  }
+}
+```
+
+Deprecation compliance can be specified for any Ember addon, including
+`ember-source`. Specifying compliance for a particular version of an addon means
+that the consuming app/addon will function without triggering any deprecations
+that were added in or prior to that version. Specifying the version number will
+turn all deprecations before the version into _assertions_ when running in DEBUG
+mode, preventing users from relying on the behavior altogether, but still
+providing a helpful error message in those cases. In production builds of
+`ember-source`, the features will be removed as possible, but they are not
+guaranteed to be removed fully.
+
+The `deprecationCompliant` field must be an exact, existing version of the
+addon. New applications and addons will have `deprecationCompliant` set to the
+latest version of `ember-source` and `ember-data` once the feature has landed.
+
+Ember's `deprecate` method will receive two new arguments that allow it to
+participate in this system:
+
+```ts
+interface DeprecationOptions {
+  id: string;
+  until: string;
+  since: string;
+  source: string;
+  url?: string;
+}
+
+declare function deprecate(message: string, test?: boolean, options?: DeprecationOptions) => void;
+```
+
+`since` must be a SemVer version equal to or lower than the current version of
+the library. `source` must be the name of the library that is creating the
+deprecation. If either of these options is missing, then the deprecation will
+not be affected by any settings for deprecation compliance in consuming
+applications. Not including these options will be deprecated and issue a
+warning.
+
+### Dependencies
+
+Apps and addons may depend on other addons that are not yet deprecation
+compliant with a given version. If the parent app or addon has a more recent
+version of a library specified, an error will be thrown, alerting the user of
+the addon that is not compliant. The user can then lower their deprecation
+compliance version to match the addon if they choose, but they _cannot_ override
+the compliance versions of their dependencies, since there is no way for us to
+know whether or not the feature is used and will cause errors otherwise.
+
+If an addon does not specify a deprecation compliance version, then it will be
```

Comment:
I don't see how this will work. Addons which specify nothing will get deprecated features stripped, but addons which specify something will block apps which have a newer version. This places a difficult burden on addon maintainers either:
 

1. continuously update this value to a new version even when no changes need to be made to the code itself.
2. leave it out and stay rigidly on top of deprecation warnings to avoid breaking apps who specifiy a new ember version

Reply:
This is called out in the drawbacks as a downside. I think automatically updating the compat version with `ember-cli-update` would help addon authors, but I'm not sure there's another way we could provide the support we're trying to, without these caveats. 

Allowing more recent versions/overrides from the consumer would just mean that the versions addon's specify has no safety at all, and the other option would be that addon's don't specify a version and we don't warn the user at all.

Any ideas?

Reply:
I don't think addons need safety here. Application authors get deprecation warnings so they should be responsible for not setting this version too high for the addons they consume. I don't think this value should be set anywhere but the top level app.

Reply:
> I don't think addons need safety here. 

I think I disagree, though only slightly. I think there are absolutely a set of addons that will want to explicitly declare their compatibility, but that most will want to leave this unspecified in their published `package.json`. Instead, addons would ensure that they do not regress on the deprecation front by specifying `ember.deprecationCompliant.ember-source` in their ember-try configuration. This ensures that the addon author is aware of the deprecations being emitted from the addon itself (during their normal workflow), and that consumers are not negatively affected (assuming the consumer app/addon's own tests pass properly).

Reply:
Then can we lean towards this RFC and resulting documentation recommend addons *not* specify this information unless they want an unhanded deprecation in their code to throw a build error in apps? Is there another case I'm not seeing?

Reply:
I have two thoughts on addons:

1) Can we use a compound key? the addon cannot declare compatibility for a version greater than the version of `ember-source` it declares in dev.

This means that addons that specify nothing wouldn't drift accidentally into non-compat over time.

2) Can we add an addon-overrides to the app's config? TL;DR the app should be able to forcibly say "x addon is compatible"